Fix version check for grad_fn for views

albanD opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#34145 Fix version check for grad_fn for views**

This fix the following behavior:
```python
import torch

class MyFn(torch.autograd.Function):
    @staticmethod
    def forward(ctx, inp, inplace):
        view = inp.clone()[:3]
        if inplace:
            view += 2
        return view

    @staticmethod
    def backward(ctx, grad):
        return grad, None

base = torch.rand(10, requires_grad=True)
foo = MyFn.apply(base, False)

print(foo.grad_fn)
# <torch.autograd.function.MyFnBackward object at 0x7f5fd28c4d18>

foo = MyFn.apply(base, True)

print(foo.grad_fn)
# <AsStridedBackward object at 0x7f601c0c3cf0>
```

Where both should be printing `MyFnBackward`.

Note: The problem here comes from the special logic of the custom autograd Function where the edge is set after multiple ops happened. As discussed with @ezyang we might want to think about a more long term solution to this as the current semantic is quite weird.

Differential Revision: [D20229907](https://our.internmc.facebook.com/intern/diff/D20229907)

albanD committed:
Fix version check for grad_fn for views

[ghstack-poisoned]

albanD committed:
Update on "Fix version check for grad_fn for views"


This fix the following behavior:
```python
import torch

class MyFn(torch.autograd.Function):
    @staticmethod
    def forward(ctx, inp, inplace):
        view = inp.clone()[:3]
        if inplace:
            view += 2
        return view

    @staticmethod
    def backward(ctx, grad):
        return grad, None

base = torch.rand(10, requires_grad=True)
foo = MyFn.apply(base, False)

print(foo.grad_fn)
# <torch.autograd.function.MyFnBackward object at 0x7f5fd28c4d18>

foo = MyFn.apply(base, True)

print(foo.grad_fn)
# <AsStridedBackward object at 0x7f601c0c3cf0>
```

Where both should be printing `MyFnBackward`.


[ghstack-poisoned]

albanD committed:
Update on "Fix version check for grad_fn for views"


This fix the following behavior:
```python
import torch

class MyFn(torch.autograd.Function):
    @staticmethod
    def forward(ctx, inp, inplace):
        view = inp.clone()[:3]
        if inplace:
            view += 2
        return view

    @staticmethod
    def backward(ctx, grad):
        return grad, None

base = torch.rand(10, requires_grad=True)
foo = MyFn.apply(base, False)

print(foo.grad_fn)
# <torch.autograd.function.MyFnBackward object at 0x7f5fd28c4d18>

foo = MyFn.apply(base, True)

print(foo.grad_fn)
# <AsStridedBackward object at 0x7f601c0c3cf0>
```

Where both should be printing `MyFnBackward`.

Differential Revision: [D20229907](https://our.internmc.facebook.com/intern/diff/D20229907)

[ghstack-poisoned]

gchanan commented on test/test_autograd.py:
```diff
@@ -3903,6 +3903,28 @@ def test_leaky_relu_inplace_with_neg_slope(self):
+        self.assertEqual(foo.grad_fn.__class__.__name__, "MyFnBackward")
+
+        foo = MyFn.apply(base, True)
+        self.assertEqual(foo.grad_fn.__class__.__name__, "MyFnBackward")
```
is the problem printing the grad function or that some behavior was wrong?

albanD replied:
The behavior was wrong: even though neither the view or the base was modified after the graph was created, we still rewrote the graph as if the inplace happened after the graph creation.

This is not ok in particular because rewriting this history for things that are views of temporary Variables is dangerous (plain wrong but for now we just warn).

ezyang commented on torch/csrc/autograd/variable.cpp:
```diff
@@ -164,6 +164,13 @@ namespace impl {
     auto* meta = materialize_autograd_meta(self);
     meta->grad_fn_ = std::move(edge.function);
     meta->output_nr_ = edge.input_nr;
+    // For views, make sure this new grad_fn_ is not overwritten unless it is necessary
```
I could believe that this is correct, but I don't see the chain of causality where setting `attr_version` prevents the `grad_fn` from being overwritten. It would be helpful to have a code pointer to the point where the overwrite happens.

albanD replied:
It is in the `VariableHooks::grad_fn()` in this same file, where we check `diff_view_meta->attr_version != current_version` to know if the graph should be rewritten or not.

albanD committed:
Update on "Fix version check for grad_fn for views"


This fix the following behavior:
```python
import torch

class MyFn(torch.autograd.Function):
    @staticmethod
    def forward(ctx, inp, inplace):
        view = inp.clone()[:3]
        if inplace:
            view += 2
        return view

    @staticmethod
    def backward(ctx, grad):
        return grad, None

base = torch.rand(10, requires_grad=True)
foo = MyFn.apply(base, False)

print(foo.grad_fn)
# <torch.autograd.function.MyFnBackward object at 0x7f5fd28c4d18>

foo = MyFn.apply(base, True)

print(foo.grad_fn)
# <AsStridedBackward object at 0x7f601c0c3cf0>
```

Where both should be printing `MyFnBackward`.

Differential Revision: [D20229907](https://our.internmc.facebook.com/intern/diff/D20229907)

[ghstack-poisoned]

ezyang reviewed (approved):
* The explanation you gave me in our VC: put it in the PR
* Make some reference to the explanation in the PR at the short comment

albanD committed:
Update on "Fix version check for grad_fn for views"


This fix the following behavior:
```python
import torch

class MyFn(torch.autograd.Function):
    @staticmethod
    def forward(ctx, inp, inplace):
        view = inp.clone()[:3]
        if inplace:
            view += 2
        return view

    @staticmethod
    def backward(ctx, grad):
        return grad, None

base = torch.rand(10, requires_grad=True)
foo = MyFn.apply(base, False)

print(foo.grad_fn)
# <torch.autograd.function.MyFnBackward object at 0x7f5fd28c4d18>

foo = MyFn.apply(base, True)

print(foo.grad_fn)
# <AsStridedBackward object at 0x7f601c0c3cf0>
```

Where both should be printing `MyFnBackward`.

Note: The problem here comes from the special logic of the custom autograd Function where the edge is set after multiple ops happened. As discussed with @ezyang we might want to think about a more long term solution to this as the current semantic is quite weird.

Differential Revision: [D20229907](https://our.internmc.facebook.com/intern/diff/D20229907)

[ghstack-poisoned]
